File: 2/solution.rb
require "../shared/read_file.rb"

input = read_file("input.txt")

# part 1

def calculate_wrapping_paper_sum(input)
  presents = parse_presents(input)
  wrapping_paper_sum = 0

  presents.each do |present|
    present_spec = parse_present_dimension(present)

    wrapping_paper_sum += calculate_wrapping_paper_per_present(present_spec)
  end

  wrapping_paper_sum
end

def calculate_wrapping_paper_per_present(spec)
  side_one = spec[:length] * spec[:width]
  side_two = spec[:width] * spec[:height]
  side_three = spec[:height] * spec[:length]
  smallest_side = [side_one, side_two, side_three].min

  (2 * side_one) + (2 * side_two) + (2 * side_three) + smallest_side
end

# part 2

def calculate_ribbon_sum(input)
  presents = parse_presents(input)
  ribbon_sum = 0

  presents.each do |present|
    present_spec = parse_present_dimension(present)

    ribbon_sum += calculate_ribbon_sum_per_present(present_spec)
  end

  ribbon_sum
end

def calculate_ribbon_sum_per_present(spec)
  sides = [spec[:length], spec[:width], spec[:height]]
  smallest_two_sides = find_smallest_two_sides(sides)
  smallest_perimeter = calculate_smallest_perimeter(smallest_two_sides)
  bow = spec[:length] * spec[:width] * spec[:height]

  smallest_perimeter + bow
end

def find_smallest_two_sides(sides)
  sorted = sides.sort do |a, b|
    a - b
  end

  [sorted[0], sorted[1]]
end

def calculate_smallest_perimeter(sides)
  side_one = sides[0]
  side_two = sides[1]

  side_one + side_one + side_two + side_two
end

# helpers

def parse_presents(present_string)
  presents = present_string.split("\n")

  presents
end

def parse_present_dimension(present)
  present_dimensions = present.split("x")

  return {
    length: present_dimensions[0].to_i,
    width: present_dimensions[1].to_i,
    height: present_dimensions[2].to_i
  }
end

puts "Part 1: #{calculate_wrapping_paper_sum(input)}"
puts "Part 2: #{calculate_ribbon_sum(input)}"

File: 2/solution.js
var init = require('../shared/init');
var FILE_NAME = 'input.txt';

init(FILE_NAME, [
  calculateWrappingPaperSum,
  calculateRibbonSum
]);

// part 1

function calculateWrappingPaperSum(input) {
  var presents = parsePresents(input);
  var wrappingPaperSum = 0;

  presents.forEach(function(present) {
    var presentSpec = parsePresentDimensions(present);

    wrappingPaperSum += calculateWrappingPaperPerPresent(presentSpec)
  });

  return wrappingPaperSum;
}

function calculateWrappingPaperPerPresent(spec) {
  var sideOne = spec.length * spec.width;
  var sideTwo = spec.width * spec.height;
  var sideThree = spec.height * spec.length;
  var smallestSide = Math.min(sideOne, sideTwo, sideThree);

  return (2 * sideOne) + (2 * sideTwo) + (2 * sideThree) + smallestSide;
}

// part 2

function calculateRibbonSum(input) {
  var presents = parsePresents(input);
  var ribbonSum = 0;

  presents.forEach(function(present) {
    var presentSpec = parsePresentDimensions(present);

    ribbonSum += calculateRibbonPerPresent(presentSpec);
  });

  return ribbonSum;
}

function calculateRibbonPerPresent(spec) {
  var sides = [spec.length, spec.width, spec.height];
  var smallestTwoSides = findSmallestTwoSides(sides);
  var smallestPerimeter = calculateSmallestPerimeter(smallestTwoSides);
  var bow = spec.length * spec.width * spec.height;

  return smallestPerimeter + bow;
}

function findSmallestTwoSides(sides) {
  var sorted = sides.sort(function(a, b) {
    return a - b;
  });

  return [sorted[0], sorted[1]];
}

function calculateSmallestPerimeter(sides) {
  var sideOne = sides[0];
  var sideTwo = sides[1];

  return sideOne + sideOne + sideTwo + sideTwo;
}

// helpers

function parsePresents(input) {
  var presents = input.trim().split('\n');

  return presents;
}

function parsePresentDimensions(present) {
  var presentDimensions = present.split('x');

  return {
    length: +presentDimensions[0],
    width: +presentDimensions[1],
    height: +presentDimensions[2]
  };
}
